federatedscope/autotune/pfedhpo/fl_server.py

```diff
@@ -0,0 +1,374 @@
+            ss = sorted([(int(k[3:]), v) for k, v in ss.items()],
+                        key=lambda x: x[0])
+            self._grid = []
+            self._cfsp = [[tp[1] for tp in ss]]
+        else:
+            # This is not a flat search space
+            # be careful for the order
+            self._grid = sorted(ss.keys())
+            self._cfsp = [ss[pn] for pn in self._grid]
+
+        super(pFedHPOFLServer, self).__init__(
+            ID, state, config, data, model, client_num,
+            total_round_num, device, strategy, **kwargs)
+        os.makedirs(self._cfg.hpo.working_folder, exist_ok=True)
+
+        if self._cfg.federate.restore_from != '':
+            if not os.path.exists(self._cfg.federate.restore_from):
+                logger.warning(f'Invalid `restore_from`:'
+                               f' {self._cfg.federate.restore_from}.')
+            else:
+                pi_ckpt_path = self._cfg.federate.restore_from[
+                               :self._cfg.federate.restore_from.rfind('.')] \
+                               + "_pfedhpo.yaml"
+                with open(pi_ckpt_path, 'r') as ips:
+                    ckpt = yaml.load(ips, Loader=yaml.FullLoader)
+                self._z = [np.asarray(z) for z in ckpt['z']]
+                self._theta = [np.exp(z) for z in self._z]
+                self._store = ckpt['store']
+                self._stop_exploration = ckpt['stop']
+                self._trace = dict()
+                self._trace['global'] = ckpt['global']
+                self._trace['refine'] = ckpt['refine']
+                self._trace['entropy'] = ckpt['entropy']
+                self._trace['mle'] = ckpt['mle']
+
+        self.train_anchor = self._cfg.hpo.pfedhpo.train_anchor
+        self.discrete = self._cfg.hpo.pfedhpo.discrete
+
+        # prepare search space and bounds
+        self._ss = parse_search_space(self._cfg.hpo.pfedhpo.ss)
+        self.dim = len(self._ss)
+        self.bounds = np.asarray([(0., 1.) for _ in self._ss])
+        self.pbounds = {}
+        if not self.discrete:
+            for k, v in self._ss.items():
+                if not (hasattr(v, 'lower') and hasattr(v, 'upper')):
+                    raise ValueError("Unsupported hyper type {}".format(type(v)))
+                else:
+                    if v.log:
+                        l, u = np.log10(v.lower), np.log10(v.upper)
+                    else:
+                        l, u = v.lower, v.upper
+                    self.pbounds[k] = (l, u)
+        else:
+            for k, v in self._ss.items():
+                if not (hasattr(v, 'lower') and hasattr(v, 'upper')):
+                    if hasattr(v, 'choices'):
+                        self.pbounds[k] = list(v.choices)
+                    else:
+                        raise ValueError("Unsupported hyper type {}".format(type(v)))
+                else:
+                    if v.log:
+                        l, u = np.log10(v.lower), np.log10(v.upper)
```

Comment:
the lower and upper bounds of `v` are always positive, right?

Reply:
yes